### PR TITLE
Begin warning on modifying global state of colormaps

### PR DESCRIPTION
**Pull request by greglucas**
## PR Summary

This is the start of deprecating global state access for built-in colormaps. It doesn't implement any solutions yet, just begins to warn users about what will happen.

Returning a copy of colormaps is implemented in #16943 

Per the dev call, the tentative plan forward is:
### 3.3
- [x] still return global
- [x] warn on set_*
- [x] make cmap_d private and warn on read/write

### 3.4:
- [ ] return a copy
- [ ] don’t warn on set_*
- [ ] warn on re-accessing a color map that would have been changed

### 3.5:
- [ ] either return a copy or make colormaps immutable

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/cm.py [hunk 1]
```diff
@@ -65,12 +65,16 @@ def _gen_cmap_d():
...
         cmap_d[rmap.name] = rmap
     return cmap_d
 
 
-cmap_d = _gen_cmap_d()
-locals().update(cmap_d)
+_cmap_d = _gen_cmap_d()
+locals().update(_cmap_d)
+# This is no longer considered public API
+cmap_d = _cmap_d
```
Comment (greglucas) [inline lib/matplotlib/cm.py:77]: I wasn't sure how to warn on read/write here. Do I need to create a new Dictionary subclass that overwrites the getters/setters with the warn_deprecated call?

Comment (anntzer) [reply lib/matplotlib/cm.py:77]: I guess something like https://github.com/matplotlib/matplotlib/pull/10735 (simpler if Py3.7+) would be needed to fully warn on access (and then you don't need to warn on getitem/setitem I guess)?

Comment (greglucas) [reply lib/matplotlib/cm.py]: That is a nice helper! I think that may be very helpful for deprecating plt.cm usage. I decided to go the quick route and just subclass dict and add in warning on get/set only.

**Comment by greglucas** [comment]
@timhoffm, just wanted to add you here to get your feedback since I can't explicitly request you as a reviewer. Happy to hear your thoughts on this.

**Comment by greglucas** [comment]
Actually, I started searching around a bit to see where/how others are using cmap_d and I think deprecating `cmap_d` needs a little more careful consideration for a good alternative/replacement. There really is no way to access a list of what all is in the cmap_d dictionary explicitly without either using cmap_d or to use the pyplot interface. We should think about adding a `get_registered_cmaps()` function or similar to list what is available to the `get_cmap()` call.

We explicitly use it in this example:
https://matplotlib.org/3.1.0/gallery/images_contours_and_fields/demo_bboximage.html

There is this SO post with the first answer stating to use `sorted(cmap_d)`:
https://stackoverflow.com/questions/47302343/what-names-can-be-used-in-plt-cm-get-cmap

There is also mention of accessing plt.cm.datad! So, the `plt.cm` registering all public methods is going to be tricky.

**Comment by QuLogic** [comment]
Maybe removal should be scheduled for 4.0 or whatever sounds like "a long time from now".

**Comment by greglucas** [comment]
@timhoffm, Did you have any thoughts on this? I'd like to try to get some of the warnings in for the 3.3 release to get the deprecation cycle started. No API code is being changed yet.

- The first commit is warning on set/get from the global cmap registry and I think everyone was in agreement about that.
- The second commit is warning on accessing cmap_d directly, which was a little harder to warn against using (Discussion above about potentially waiting until 3.7 only for a dependency).

File: lib/matplotlib/colors.py [hunk 2]
```diff
@@ -484,6 +484,21 @@ def makeMappingArray(N, data, gamma=1.0):
     return _create_lookup_table(N, data, gamma)
 
 
+def _deprecate_global_cmap(cmap):
+    if hasattr(cmap, '_global') and cmap._global:
```
Comment (timhoffm) [inline lib/matplotlib/colors.py:488]:
```suggestion
    if getattr(cmap, '_global', False):
```

File: lib/matplotlib/colors.py [hunk 3]
```diff
@@ -484,6 +484,21 @@ def makeMappingArray(N, data, gamma=1.0):
     return _create_lookup_table(N, data, gamma)
 
 
+def _deprecate_global_cmap(cmap):
```
Comment (timhoffm) [inline lib/matplotlib/colors.py:487]: A bit more precise:
```suggestion
def _warn_if_global_cmap_modified(cmap):
```
The main purpose here is to warn. Technically, the deprecation has already happend.

File: lib/matplotlib/colors.py [hunk 4]
```diff
@@ -599,13 +614,15 @@ def __copy__(self):
...
             cmapobject._lut = np.copy(self._lut)
+        cmapobject._global = False
         return cmapobject
 
     def set_bad(self, color='k', alpha=None):
         """Set the color for masked values."""
         self._rgba_bad = to_rgba(color, alpha)
         if self._isinit:
             self._set_extremes()
+        _deprecate_global_cmap(self)
```
Comment (timhoffm) [inline lib/matplotlib/colors.py:625]: I would put such warnings in the first line of the method, before any related code is executed. Same for `set_under`, `set_over`.

File: lib/matplotlib/cm.py [hunk 5]
```diff
@@ -130,9 +136,12 @@ def get_cmap(name=None, lut=None):
     Parameters
     ----------
     name : `matplotlib.colors.Colormap` or str or None, default: None
-        If a `.Colormap` instance, it will be returned.  Otherwise, the name of
-        a colormap known to Matplotlib, which will be resampled by *lut*.  The
-        default, None, means :rc:`image.cmap`.
+        If a `.Colormap` instance, it will be returned.
+        Otherwise, the name of a colormap known to Matplotlib, which will
+        be resampled by *lut*. Currently, this returns the global colormap
```
Comment (timhoffm) [inline lib/matplotlib/cm.py:164]: I'd make the change announcement a separate paragraph to give it more weight.

File: lib/matplotlib/cm.py [hunk 6]
```diff
@@ -65,12 +65,39 @@ def _gen_cmap_d():
...
         rmap = cmap.reversed()
+        cmap._global = True
+        rmap._global = True
         cmap_d[rmap.name] = rmap
     return cmap_d
 
 
-cmap_d = _gen_cmap_d()
-locals().update(cmap_d)
+class _DeprecatedCmapDict(dict):
```
Comment (timhoffm) [inline lib/matplotlib/cm.py:74]: Deriving from dict makes `_DeprecatedCmapDict` hold its own data.

` cmap_d = _DeprecatedCmapDict(_cmap_d)` below will make a copy, and both instances will run out of sync when the user starts to modify `cmap_d`.

I recommend deriving `_DeprecatedCmapDict` from `MutableMapping`, holding `_cmap_d` as a private attribute and forwarding all calls to `self._cmap_d`.

**Comment by timhoffm** [comment]
:+1:  for pushing the deprecations into 3.3.

**Comment by greglucas** [comment]
Thanks, Tim! Great suggestions. I just pushed up a commit with all of those suggestions in it.

I also decided to take your advice for more descriptive names and apply it to the private `_cmap_d` that I added in. It is now called `_cmap_registry`. I can change that back if desired, but figure this will help to describe what the dictionary actually is.

File: lib/matplotlib/cm.py [hunk 7]
```diff
@@ -65,12 +66,53 @@ def _gen_cmap_d():
...
-cmap_d = _gen_cmap_d()
-locals().update(cmap_d)
+class _DeprecatedCmapDict(MutableMapping):
+    """Dictionary mapping for deprecated _cmap_d access."""
+
+    def __init__(self, cmap_registry):
+        self._cmap_registry = cmap_registry
+
+    def __delitem__(self, key):
+        self._cmap_registry.__delitem__(key)
```
Comment (timhoffm) [inline lib/matplotlib/cm.py:82]: You can also warn on deletion.

File: lib/matplotlib/cm.py [hunk 8]
```diff
@@ -104,7 +146,8 @@ def register_cmap(name=None, cmap=None, data=None, lut=None):
             raise ValueError("Arguments must include a name or a "
                              "Colormap") from err
     if isinstance(cmap, colors.Colormap):
-        cmap_d[name] = cmap
+        cmap._global = True
```
Comment (timhoffm) [inline lib/matplotlib/cm.py:149]: Please document in the docstring that if passing a Colormap instance, that instance will become immutable.

Comment (greglucas) [reply lib/matplotlib/cm.py:157]: I added in a Notes section to both the `get_cmap` and `register_cmap` functions now. Let me know if you want any different wording there. It may be odd to talk about future behavior there considering you can't do anything to get that future behavior yet? But I can't think of any great wording off the top of my head.

File: lib/matplotlib/cm.py [hunk 9]
```diff
@@ -65,12 +66,54 @@ def _gen_cmap_d():
...
         rmap = cmap.reversed()
+        cmap._global = True
+        rmap._global = True
         cmap_d[rmap.name] = rmap
     return cmap_d
 
 
-cmap_d = _gen_cmap_d()
-locals().update(cmap_d)
+class _DeprecatedCmapDict(MutableMapping):
```
Comment (tacaswell) [inline lib/matplotlib/cm.py:75]: Very minor naming quibble, can we put the word "Wrapper" in this class?  It took me a bit of thought to understand that this is holding a reference to `_cmap_registry` rather than having its own state.

Feel free to disregard this comment if you don't think it makes things clearer.

File: lib/matplotlib/colors.py [hunk 10]
```diff
@@ -484,6 +484,21 @@ def makeMappingArray(N, data, gamma=1.0):
...
+        cbook.warn_deprecated(
+            "3.3",
+            message="You are modifying the state of a globally registered "
+                    "colormap. In future versions, you will not be able "
+                    "to modify a globally registered colormap directly. "
+                    "To eliminate this warning until then, you can make "
+                    "a copy of the requested colormap before modifying it. ",
+            alternative="To modify a colormap without overwriting the "
+                        "global state, you can make a copy of the colormap "
+                        f"first. cmap = copy.copy(mpl.cm.get_cmap({cmap.name})"
```
Comment (tacaswell) [inline lib/matplotlib/colors.py:498]: This is a pretty verbose alternative for something that I expect to catch a fair number of users.  I am in favor of adding the `copy=True` kwarg to `get_cmap` in the same move as we add this warning so we can provide a less verbose (if less back-compatible) option to users.

**Comment by tacaswell** [comment]
I labeled as release critical as I think this is very close and do not want it to fall through the cracks as we do the final push to 3.3.

**Comment by greglucas** [comment]
👍 
If anyone has ideas on why the docs are failing, let me know... I tried to look for what the Error was and couldn't figure out how to parse the Sphinx output for it. It must be something with docstring formatting?

Good comments on the rest of it. I'll try and push up changes to address them tomorrow.

**Comment by tacaswell** [comment]
Can you try rebasing on current master?  

We are very cranky about making sure all of the internal references actually work (with a list of allowed errors).  Sphinx made some internal changes that affected how references to the nested classes work so we re-generated the allowed list to the new version.  Unlike travis which runs against the shadow merge commit GH makes to determine if your branch is mergable, circle runs against your actually commit so if you have not rebased on master from when we fixed that circle will be broken.

File: lib/matplotlib/cm.py (hunk 9 above)
Comment (greglucas) [reply lib/matplotlib/cm.py]: No problem. This shouldn't really be in the code base for very long anyways.

File: lib/matplotlib/colors.py (hunk 10 above)
Comment (greglucas) [reply lib/matplotlib/colors.py]: I decided to just change the wording here for now, and put `.copy()` instead of `copy.copy()`. I think the wording was just a little long, so I tidied it up. The alternative is actually less verbose than `copy=True`. I also don't really know if we should add a keyword for one version that will just be removed a version later when it is unnecessary. Let me know what you think...

File: lib/matplotlib/tests/test_colors.py [hunk 11]
```diff
@@ -70,6 +70,73 @@ def test_register_cmap():
         cm.register_cmap()
 
 
+@pytest.mark.skipif(matplotlib.__version__ < "3.5",
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_colors.py:73]: I'm hesitant adding tests for future intended behavior. We don't know for sure if this will be implemented for the version and with the behavior we anticipate now.

Additionally, the test will be ignored during the complete development phase (master version targeting 3.3 is currently evaluated to `3.2.1.post2103+gb94812c60`). The check would only hit very late during the release.

You may keep this, but it needs explanation. If we are hit by the failing test in a year or so, we need a clear way of finding out why this is here, why the tested functionality is desired and what we have to do to implement it. Therefore, please add a comment describing the intention and referring to the modification roadmap discussed on github.

Comment (greglucas) [reply lib/matplotlib/tests/test_colors.py:73]: Yeah, that is unfortunate that this wouldn't actually hit until near the release. Is there a way to test against something else? My thought is that leaving these there will make sure this doesn't get lost along the way and it isn't hard to implement these since I've already got the code in another linked PR. I put a link to the gh issue number in as a comment now.

File: examples/images_contours_and_fields/demo_bboximage.py [hunk 12]
```diff
@@ -38,7 +38,7 @@
 a = np.vstack((a, a))
 
 # List of all colormaps; skip reversed colormaps.
-maps = sorted(m for m in plt.cm.cmap_d if not m.endswith("_r"))
+maps = sorted(m for m in plt.colormaps() if not m.endswith("_r"))
```
[the remark this answers is not in the document]
Comment (greglucas) [reply examples/images_contours_and_fields/demo_bboximage.py:41]: I had the same thought: https://github.com/matplotlib/matplotlib/pull/16991#issuecomment-607548990

I'm not sure what the best step is to expose the access, but would agree that I don't want to have to go to pyplot if not necessary. Maybe rather than a `cmap_d`, we could do something like your MutableMapping suggestion and then override the get/set methods to ensure copy/immutable return.

[the remark this answers is not in the document]
Comment (timhoffm) [reply examples/images_contours_and_fields/demo_bboximage.py:41]: *Sorry if this is a bit unstructured. I just stated typing and ideas unfolded along the way.*

There would be two ways:

1. The `cm` module currently uses simple functions `get_cmap()` and `register_cmap()`. In that spirit, you could add a function `get_cmap_names()` (naming t.b.d.) but we have to make a distinction between `get_cmap` returning an instance and that function (probably?) returning strings.

2. The alternative would be to expose a `ColormapRegistry` object, which provides all necessary functionality. That would likely be subclassing `MutableMapping` to look dict-like but still perform all relevant checks (maybe it would alternatively only implement `Mapping` and provide an explicit register method if we feel the mutation should be more special (i.e. we don't want `del registry['jet']` or `registry['jet'] = Colormap(...)`).
This would also replace the module level functions; at least their implementation. Whether they will additionally stay for backward compatibility is a separate topic.

In the end 2 is the more complex approach. However, I have a slight preference for it because it properly encapsulates the colormap handling. On the downside it changes the API from calling simple cm functions to an object. But then again, we could move the actual instance to a variable `matplotlib.colormaps`, which may make more sense and users would usually not need to import matplotlib.cm anymore; and `pyplot.colormaps` could later even point to the same instance (with some API compatibility because the keys of that mapping would then colormap names, so `for name in plt.colormaps()` would still work.

File: lib/matplotlib/tests/test_colors.py (hunk 11 above)
Comment (timhoffm) [reply lib/matplotlib/tests/test_colors.py:73]: Other tests could be earlier but would still hit at some other time and break CI, when we maybe don't want to implement this right now.

Either leave it like this or remove it here and put the test just into your PR.

File: examples/images_contours_and_fields/demo_bboximage.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (greglucas) [reply examples/images_contours_and_fields/demo_bboximage.py:41]: I agree with pretty much everything you're saying. I also don't think 2 would be that hard to implement. An issue with naming will be found with: #11413 which is already proposing to rename cm to colormaps. This discussion is really more a part of the MEP: https://matplotlib.org/3.2.1/devel/MEP/MEP21.html
Maybe we should do some larger work on an overarching vision there and then decide on how best to get there?

[the remark this answers is not in the document]
Comment (timhoffm) [reply examples/images_contours_and_fields/demo_bboximage.py:41]: If we provide colormap registry access from `matplotlib` directly, `cm.get_cmap` and `cm.register_cmap` become obsolete. There won't be a need for the end user to import `cm`, which in turn reduces the need for renaming (#11413). But we could still find other names for that.

It would be reasonable to detail the overall vision.

File: lib/matplotlib/cm.py [hunk 13]
```diff
@@ -65,12 +66,54 @@ def _gen_cmap_d():
...
+
+    def __delitem__(self, key):
+        self._warn_deprecated()
+        self._cmap_registry.__delitem__(key)
+
+    def __getitem__(self, key):
+        self._warn_deprecated()
+        return self._cmap_registry.__getitem__(key)
+
+    def __iter__(self):
```
Comment (tacaswell) [inline lib/matplotlib/cm.py:89]: This does not warn for `list(mpl.cm.cmap_d)` but does warn for `list(mlp.cm.cmap_d.items())`.   Was this an oversight that `__iter__` does not warn or a decision that because it is only giving a list of keys that could then go to `get_cmap` it should not wark?

File: lib/matplotlib/tests/test_colors.py (hunk 11 above)
Comment (tacaswell) [reply lib/matplotlib/tests/test_colors.py:73]: It is better to put the test in the PR where the change is actually made.

There are better ways to document a future test than a skipped test (I don't think this would actually get run until we tagged the rc.  

As a side note, In general comparing versions via text is not the best way as it sorts pre-releases as being later than the final releases, `distuitls.version.LooseVersion` is usually the better option.

File: lib/matplotlib/cm.py (hunk 13 above)
Comment (greglucas) [reply lib/matplotlib/cm.py:89]: Oversight, the iter still warns because of the getitem, but I didn't think of people just listing the colormaps and using it manually themselves elsewhere. I just added it here as well.

File: lib/matplotlib/tests/test_colors.py (hunk 11 above)
Comment (greglucas) [reply lib/matplotlib/tests/test_colors.py]: Thanks for that suggestion of comparison with distutils. I just removed all of these future tests. I'll hope it doesn't drop off my radar.

File: lib/matplotlib/cm.py (hunk 13 above)
Comment (tacaswell) [reply lib/matplotlib/cm.py:89]: same for `__len__` below?   Sorry for the thousand paper cuts at the end...

Comment (greglucas) [reply lib/matplotlib/cm.py:89]: No worries, easy enough to fix. Should have just done it myself ;)

**Comment by mperrin** [comment]
Just wanted to say, bravo for the super clear deprecation message including specific code suggestions for how to update downstream code. I can see that took extra effort to put together and it's much appreciated.  👏

**Comment by jklymak** [comment]
I wish I'd followed this better.  I did

```python
cmap = cm.RdBu_r
cmap.set_over('yellow')
```
and was told I had to do:

```
MatplotlibDeprecationWarning: You are modifying the state of a globally registered colormap. In future versions, you will not be able to modify a registered colormap in-place. To remove this warning, you can make a copy of the colormap first. cmap = copy.copy(mpl.cm.get_cmap("RdBu_r"))
```

Thats pretty user hostile.  When will this get fixed?

**Comment by greglucas** [comment]
I guess it depends on what you mean by fixed.
#18503 will return a copy of the colormap, which should remove a lot of the Deprecation dance. That will hopefully be in 3.4...

We could update the message pretty easily too if you have any suggestions there.

**Comment by jklymak** [comment]
I guess I mean an alternative that doesn't require an import and copy call.  If that is #18503 we should get that in ASAP.  

I don't understand why a deprecation was deemed necessary for that PR to go in though.  I'd actually vote to revert this for V3.3.3, but maybe I'm not understanding some subtlety.

**Comment by greglucas** [comment]
This was to enable people who may have been (ab)using the ability to modify the global colormap to change before we remove the capability entirely.

In your example, if you keep running the program, or go to some other Jupyter notebook cell and want to get the RdBu colormap again, it would still have the yellow over color, which is probably not desired.
```python
cmap = cm.RdBu_r
cmap.set_over('yellow')

# Later on you get the built-in colormap again
cmap2 = cm.get_cmap('RdBu_r')
cmap2(np.inf)  # yellow
```

I agree it is somewhat annoying to import copy for now. We had discussed bringing in a `copy=True` keyword argument, but then we'd just have to deprecate that after a single release once we always return a copy/immutable version of the colormap.

**Comment by jklymak** [comment]
Sure, I agree with the over all strategy, just not the tactics.  As of 3.3 there is no other way to get the colormap except to do `copy.copy`.  So now our users who need this, change to copy.copy.  And then in 3.4 we offer a way to return the colormap, so now those users have to change again.   Why didn't we just introduce the change?

**Comment by greglucas** [comment]
> As of 3.3 there is no other way to get the colormap except to do copy.copy

Maybe this is just a poorly worded message then. The only reason to use copy here is if you need to re-get the original RdBu colormap at some later time. You can certainly keep using the `cm.RdBu_r` throughout your code as you normally would, and the warnings should only happen once when you are setting the over/under/bad on a built-in colormap.

My vote would be to leave this as is for now and make sure 3.4 is returning a copy. I really don't think this is that egregious, more of a minor annoyance along the deprecation path.

**Comment by anntzer** [comment]
fwiw I guess https://github.com/matplotlib/matplotlib/pull/14645 may help too?

**Comment by jklymak** [comment]
The message would be useful if people had to change their code.  But so far as I can tell we are going to allow `cmap=cm.RdBu_r` will in the future return a copy, so there is no code we need to warn the user to change.  

The only users who would have to change behaviour are those who were for some reason depending on being able to access the colormap from the global state.
```python
cmap = cm.RdBu
cmap.set_over_color([0, 1., 1., 1.])

.....
cmap=cm.RdBu
assert cmap.get_over_color() == [0, 1., 1., .1.]
```
which I imagine is exceedingly rare.  The rest get a misleading Deprecation message that makes it sound like they won't be able to call `set_over_color` on the object returns by `cm.RdBu` any more.

**Comment by tacaswell** [comment]
Did we decide to go down the route of module properties (which is py37 so we have it on main)? I do not remember any discussion of it and am not finding it in the text of this issue.

If we do not, then we can't make `cmap = cm.RdBu` to return a copy as it is Python assignment and we don't have a hook to make the copy.

**Comment by greglucas** [comment]
Good question. I think that was brought up as a possibility and an alternative would be to make the cmap object's immutable (Tim's proposal). I think we could get away with something like defining our own getattr at the module level for the "return a copy" case:

```python
def __getattr__(name):
    return ColormapRegistry[name]
```

All of this discussion leads me to think we should really have a clear end-goal in mind and lay that out. Then, we can start shimming the deprecations in. There is already [MEP21](https://matplotlib.org/devel/MEP/MEP21.html), which looks to be stagnant. Mind if I revive that MEP with some new text and then we can start linking issues/fixes to a more central location?

**Comment by jklymak** [comment]
... or use the Projects feature...

**Comment by anntzer** [comment]
Making attribute access return new (mutable) copies seems rather contrary to standard python semantics.  In particular this means that something like
```python
from matplotlib import cm
cm.jet.set_bad(...)
plt.imshow(..., cmap=cm.jet)
```
would silently "fail", whereas
```python
from matplotlib.cm import jet
jet.set_bad(...)
plt.imshow(..., cmap=jet)
```
would succeed... but I doubt that having both snippets behave differently is a good idea.

**Comment by tacaswell** [comment]
I agree with @anntzer , while cute there are too many weird cases.  Just because you _can_ hijack and change something in Python does not mean you _should_.

We need to decide if we only want to deprecate mutating the global state when getting the colormap through `get_cmap` and let people continue to mutate global state with e.g. `cm.viridis` (which then means we will have to decide if `get_cmap('foo')` returns a fresh copy of the default state (which would make `cm.viridis` and `cm.get_cmap('viridis') decoupled) or a copy of the currently accessible one at the module level (which would defeat the purpose of trying to protect the user from spooky action-at-a-distance)) or make the attribute access colormaps actually immutable (which will break a bunch of people).

I think when we wrote the warning we were not thinking about the attribute access case at all.  Do we document / recommend that usage anywhere?

**Comment by jklymak** [comment]
I find it useful in these cases to think about what the *ideal* situation would be and only then figure out how we would get there from where we are (if possible).  So, how would this look if we were to start over?

**Comment by anntzer** [comment]
My 50c would be that we should just make all colormaps immutable (after merging the with_extremes PR (or similar functionality) and possibly with a very long deprecation period, given that we are basically breaking *everyone* who way setting over/under/bad before with essentially no way to write code that compatible both with old and new Matplotlib versions, as with_extremes did not exist before).

Especially if we are going to push in the direction of encouraging the use of the global cmap registry (via `register_cmap`, which I didn't particularly like but I guess that discussion is closed), any colormap can potentially become global state, e.g. via
```python
# in foo.py
foo_cmap = LinearSegmentedColormap(...)
register_cmap("foo", foo_cmap)
...
# in module2.py
imshow(..., cmap="foo")
...
# in module3.py
from foo import foo_cmap
foo_cmap.set_over(...)  # oops, this will affect module2.py if run before it.
```

**Comment by jklymak** [comment]
It seems to me that the original problem was defining globals like `cm.jet` in the first place.  If we had just stuck with `jet=cm.get_cmap('jet')` then there would not be a problem.  

Given the state we are in, it seems sensible to me that if you modify a global like `cm.jet` (i.e. `cm.jet.set_over`) you get what you asked for, and `cm.jet` is changed everywhere.  However, if you call `jet=cm.get_cmap('jet')` you should get a local copy that does not affect the global state.  

It would be nice if `jet = cm.jet` *also* returned a copy.  Is it even possible to make that return a copy?

TLDR: it would be nice if we could return copies whenever possible, because I don't think it is at all clear to users that `cm.jet` is a global.

**Comment by anntzer** [comment]
Actually, now that I look at it again, I guess a solution for the "sneaky global" problem I mentioned just above would be for `register_cmap` to be in charge of registering *a copy of* the colormap, so that later changes to the colormap passed as parameter to `register_cmap` don't affect the global registry.  I think this is similar to how when you do `y = [...]; plt.plot(y)`, later changes to `y` don't affect the plotted line.

**Comment by timhoffm** [comment]
@anntzer it's rarely the problem that one registers a colormap instance and then modifies the instance, thus modifying the registered colormap. While that's a theoretical issue I'm not seriously worried about that. Few people register colormaps themselves, and even fewer will hold on to instances some times and use the registry other times.

The far more common problem is obtaining a reference to the registered colormap and modifying that, without knowing that it affects the registered state.

**Comment by anntzer** [comment]
I'd guess making a copy at registration time is simple enough to be worth considering, but won't insist on that either.

**Comment by tacaswell** [comment]
The module attribute access has been with us from at least 2004 (e34a333d00814124d3e19d462b9d78ac35e7a49a).  I think clawing those back is off the table. 

So if we were to make the attribute access return a copy:

 Pro: same behavior as `get_cmap`
 Con: very surprising, leads to cases where the code _looks_ right, but does not do what it expects

I think the case of

```
cm.viridis.set_under('w')  # expect to mutate in place
ax.imshow(..., cmap=cm.viridis)  # does not work!!
```

is enough to take that option out of consideration (even if we could make it work)


If we back off making the attribute access immutable:

Pro: leave the foot-cannon that people are using in place
Con: but the foot-cannon will only half work as things like

```
cm.jet.set_under('k')
ax.imshow(..., 'jet')
```
which is a variation on the issue with making `cm.jet` return a copy.  The things that make me angriest when using APIs are when it _runs_, does not do what I want, but the most "obvious" (and yes I know "obvious" is super subjective) mental model of how it works inside says it _should_ do what I want.

It also will make `cm.<name>` and `cm.get_cmap(name)` give possibly different results which seems too confusing to do

If we make the attribute access immutable (which is the status quo)

Pro: avoids all of the cons above
Con: will require a wider net of changes than we discussed going into this.

Backing off of making the global versions of the cmap objects immutable is also an option, but I think that the action-at-a-distance and general global state issues are worth cleaning up for all of the reasons we had when we started this.

----

On net, I think we should go with making the attribute access immutable and not try to return a copy.  It may not be ideal, but I think it is the least-bad of the options we have. 

----

I agree about the copy in cmap_register (I checked, we do not currently do that but we do flag them as global).

**Comment by timhoffm** [comment]
It!'s reasonable to do. I'm just clarifying that it does not solve our main issue.

**Comment by jklymak** [comment]
At some point @efiring wanted to change the name of this module anyhow.  Is yet another possibility just writing a new way to do this from scratch and leaving `cm` the way it is, and  soft-deprecating it?
